apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClient.cs

```diff
@@ -509,42 +662,43 @@ private void ShutdownIfCompleted()
             }
         }
 
-        private void DelaySendCurrentBatch()
+        private void DelaySendBatch(KeyState state, Batch batch)
         {
-            // Pre-condition: Must be locked
+            // Pre-condition: Must be locked.
             if (_batchDelayThreshold is TimeSpan batchDelayThreshold)
             {
-                // Read cancellation token here, in case the current batch changes before the task below starts.
-                var currentBatchToken = _currentBatch.TimerCts.Token;
-                // Ignore result of this Task. If it's cancelled, it's because the batch has already been sent.
                 _taskHelper.Run(async () =>
                 {
-                    using (var timerCancellation = CancellationTokenSource.CreateLinkedTokenSource(currentBatchToken, _hardStopCts.Token))
+                    using (var timerCancellation = CancellationTokenSource.CreateLinkedTokenSource(batch.ProcessedCts.Token, _hardStopCts.Token))
                     {
-                        await _taskHelper.ConfigureAwaitHideCancellation(() => _scheduler.Delay(batchDelayThreshold, timerCancellation.Token));
-                        // If batch has already moved to queue, timerToken will have been cancelled.
-                        lock (_lock)
+                        var cancelled = await _taskHelper.ConfigureAwaitHideCancellation(() => _scheduler.Delay(batchDelayThreshold, timerCancellation.Token));
+                        if (!cancelled)
                         {
-                            // Check for cancellation inside lock to avoid race-condition.
-                            if (!timerCancellation.IsCancellationRequested)
+                            // If batch has already been processed, ProcessedCts will have been cancelled.
+                            lock (_lock)
                             {
-                                // Force queuing of the current batch, whatever the size.
-                                // There will always be at least one message in the batch.
-                                QueueCurrentBatch();
+                                // Check for cancellation inside lock to avoid race-condition.
```

Comment:
This comment worries me given the part earlier about cancellation happening outside the lock...

Reply:
The previous cancellation is a `Task`., which when completed may run arbitrary user code.
This cancellation is a cancellation-token which will only be running code within this file. Cancellations of this token may occur when `_lock` is locked, but `Monitor` is re-entrant so it's OK if this continuation executes synchronously.